File: packages/east-py-datascience/src/east_py_datascience/optimization/optimization.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""Iterative coordinate descent optimization for East.

Provides discrete combinatorial optimization by iteratively optimizing each
element of a parameter vector over its candidate values. Supports multi-start
sampling for better exploration of the search space.

Ported from the Julia IterativeDecisionAlgorithm (ArrayParameterSpace branch).
"""

import random
from collections.abc import Callable
from typing import Any

import numpy as np
from east.runtime.platform import PlatformFunction
from east.types.types import (
    ArrayType,
    BooleanType,
    FloatType,
    FunctionType,
    IntegerType,
    NullType,
    OptionType,
    StructType,
    VariantType,
    VectorType,
)
from east.types.values import EastArray, EastStruct, EastVariant, EastVector, is_east_variant

# ============================================================================
# Type Definitions (must match TypeScript exactly)
# ============================================================================

InitialStrategyType = VariantType(
    [
        ("first", NullType),
        ("random", NullType),
    ]
)

EvaluationOrderType = VariantType(
    [
        ("sequential", NullType),
        ("random", NullType),
    ]
)

IterativeConfigType = StructType(
    [
        ("iterations", OptionType(IntegerType)),
        ("samples", OptionType(IntegerType)),
        ("initial", OptionType(InitialStrategyType)),
        ("order", OptionType(EvaluationOrderType)),
        ("random_state", OptionType(IntegerType)),
    ]
)


IterativeResultType = StructType(
    [
        ("best_parameters", VectorType(IntegerType)),
        ("best_objective", FloatType),
        ("iterations", IntegerType),
        ("evaluations", IntegerType),
        ("success", BooleanType),
    ]
)


# ============================================================================
# Helper Functions
# ============================================================================


def _get_option(opt: EastVariant | None, default: Any) -> Any:
    """Extract value from Option variant, returning default if None.

    Note: The runtime creates EastVariant instances, not EastOption instances,
    even for Option types. So we check the tag directly rather than using
    is_east_option().
    """
    if opt is None:
        return default
    if is_east_variant(opt) and opt.type == "some":
        return opt.value
    return default


# ============================================================================
# Platform Function Implementation
# ============================================================================


def optimization_iterative_impl(
    objective_fn: Callable[[EastVector], float],
    parameter_spaces: EastArray,
    config: EastStruct,
) -> EastStruct:
    """Run iterative coordinate descent optimization.

    Maximizes an objective function over a vector of discrete parameters.
    Each parameter position has its own set of candidate values (vector).
    The algorithm performs coordinate descent: for each element, try all
    candidate values while holding others fixed, keep the best.

    Multiple independent restarts (samples) improve exploration.

    Args:
        objective_fn: Objective function (Vector<V> -> Float), higher is better
        parameter_spaces: Per-element candidate values (Array<Vector<V>>)
        config: Optimization configuration

    Returns:
        EastStruct with best_parameters, best_objective, iterations, evaluations, success
    """
    # Extract config with defaults
    max_iterations = int(_get_option(config.get("iterations"), 100))
    num_samples = int(_get_option(config.get("samples"), 1))

    initial_opt = _get_option(config.get("initial"), None)
    use_random_init = (
        initial_opt is not None
        and is_east_variant(initial_opt)
        and initial_opt.type == "random"
    )

    order_opt = _get_option(config.get("order"), None)
    use_random_order = (
        order_opt is not None
        and is_east_variant(order_opt)
        and order_opt.type == "random"
    )

    seed = _get_option(config.get("random_state"), None)
    if seed is not None:
        seed = int(seed)

    rng = random.Random(seed)
    n = len(parameter_spaces)

    # Determine element type and numpy dtype from the first space
    if n > 0:
        space0: EastVector = parameter_spaces[0]
        element_type = space0.element_type
        dtype = space0.data.dtype
    else:
        element_type = IntegerType
        dtype = np.dtype(np.int64)

    global_best_obj = float("-inf")
    global_best_params: EastVector | None = None
    total_iterations = 0
    total_evaluations = 0

    for _sample in range(num_samples):
        # Initialize parameters as numpy vector
        if use_random_init:
            init_values = np.array(
                [rng.choice(space.data.tolist()) for space in parameter_spaces],
                dtype=dtype,
            )
        else:
            # first: use the first candidate value from each space
            init_values = np.array(
                [space.data[0] for space in parameter_spaces],
                dtype=dtype,
            )

        params = EastVector(element_type, data=init_values)

        # Evaluate initial solution
        best_obj = float(objective_fn(params))
        best_params = EastVector(element_type, data=params.data.copy())
        total_evaluations += 1

        # Coordinate descent loop
        for _iteration in range(1, max_iterations + 1):
            changed = False

            for i in range(n):
                space: EastVector = parameter_spaces[i]
                candidates = space.data.tolist()
                if use_random_order:
                    rng.shuffle(candidates)

                current_best_val = params.data[i]

                for candidate in candidates:
                    params.data[i] = candidate
                    obj = float(objective_fn(params))
                    total_evaluations += 1

                    if obj > best_obj:
                        best_obj = obj
                        best_params = EastVector(element_type, data=params.data.copy())
                        current_best_val = candidate
                        changed = True

                # Restore best value for this element
                params.data[i] = current_best_val

            total_iterations += 1

            if not changed:
                break

        # Update global best across samples
        if best_obj > global_best_obj:
            global_best_obj = best_obj
            global_best_params = best_params

    return EastStruct(
        {
            "best_parameters": global_best_params
            if global_best_params is not None
            else EastVector(element_type, data=np.array([], dtype=dtype)),
            "best_objective": global_best_obj
            if global_best_params is not None
            else 0.0,
            "iterations": total_iterations,
            "evaluations": total_evaluations,
            "success": global_best_params is not None,
        }
    )


# ============================================================================
# Platform Function Registration
# ============================================================================

optimization_impl = [
    PlatformFunction(
        name="optimization_iterative",
        inputs=[
            FunctionType([VectorType(IntegerType)], FloatType),
            ArrayType(VectorType(IntegerType)),
            IterativeConfigType,
        ],
        output=IterativeResultType,
        type="sync",
        fn=optimization_iterative_impl,
    ),
]


__all__ = [
    # Platform implementation
    "optimization_impl",
    # Types
    "InitialStrategyType",
    "EvaluationOrderType",
    "IterativeConfigType",
    "IterativeResultType",
]

File: packages/east-py-datascience/src/optimization/optimization.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */

/**
 * Iterative coordinate descent optimization for discrete combinatorial problems.
 *
 * Provides element-wise optimization over arrays of discrete values.
 * Each element is independently optimized by trying all candidate values
 * while holding other elements fixed. Multi-start sampling improves
 * exploration of the search space.
 *
 * Ported from the Julia IterativeDecisionAlgorithm (ArrayParameterSpace branch).
 *
 * @packageDocumentation
 */

import {
    East,
    StructType,
    VariantType,
    OptionType,
    ArrayType,
    IntegerType,
    BooleanType,
    FloatType,
    NullType,
    FunctionType,
} from "@elaraai/east";

// ============================================================================
// Type Definitions
// ============================================================================

/**
 * Initial value strategy for parameters.
 *
 * - `first`: Use the first candidate value from each space
 * - `random`: Randomly select from each space
 */
export const InitialStrategyType = VariantType({
    first: NullType,
    random: NullType,
});

/**
 * Evaluation order for candidate values within each element's space.
 *
 * - `sequential`: Try candidates in the order they appear
 * - `random`: Shuffle candidates before trying
 */
export const EvaluationOrderType = VariantType({
    sequential: NullType,
    random: NullType,
});

/**
 * Configuration for iterative optimization.
 *
 * All fields are optional with sensible defaults.
 */
export const IterativeConfigType = StructType({
    /** Maximum coordinate descent iterations per sample (default: 100) */
    iterations: OptionType(IntegerType),
    /** Number of independent restarts (default: 1) */
    samples: OptionType(IntegerType),
    /** How to initialize parameter values (default: first) */
    initial: OptionType(InitialStrategyType),
    /** Order to evaluate candidates (default: sequential) */
    order: OptionType(EvaluationOrderType),
    /** Random seed for reproducibility */
    random_state: OptionType(IntegerType),
});

/**
 * Result of iterative optimization.
 *
 * Generic over V (the parameter value type).
 */
export const IterativeResultType = StructType({
    /** Best parameter values found */
    best_parameters: ArrayType("V"),
    /** Objective value at best parameters */
    best_objective: FloatType,
    /** Total coordinate descent iterations across all samples */
    iterations: IntegerType,
    /** Total number of objective evaluations */
    evaluations: IntegerType,
    /** Whether optimization succeeded */
    success: BooleanType,
});

// ============================================================================
// Platform Functions
// ============================================================================

/**
 * Iterative coordinate descent optimization (generic over value type V).
 *
 * Maximizes an objective function over an array of discrete parameters.
 * Each parameter position has its own set of candidate values.
 * The algorithm optimizes one element at a time (coordinate descent),
 * with multiple independent restarts (samples).
 *
 * @example
 * ```ts
 * import { East, ArrayType, IntegerType, FloatType, variant } from "@elaraai/east";
 * import { Optimization } from "@elaraai/east-py-datascience";
 *
 * // Objective: maximize sum of parameter values
 * const objective = East.function([ArrayType(IntegerType)], FloatType, ($, params) => {
 *     const total = $.let(East.value(0.0));
 *     $.for(East.value(0n), params.length(), ($, i) => {
 *         $(total.set(total.get().add(params.get(i).toFloat())));
 *     });
 *     return $.return(total.get());
 * });
 *
 * const spaces = $.let([[0n, 1n, 2n], [0n, 1n, 2n], [0n, 1n, 2n]]);
 * const config = $.let({
 *     iterations: variant('some', 10n),
 *     samples: variant('some', 3n),
 *     initial: variant('some', variant('random', null)),
 *     order: variant('some', variant('sequential', null)),
 *     random_state: variant('some', 42n),
 * });
 *
 * const result = $.let(Optimization.iterative([IntegerType], objective, spaces, config));
 * // result.best_parameters = [2n, 2n, 2n], result.best_objective = 6.0
 * ```
 */
export const optimization_iterative = East.genericPlatform(
    "optimization_iterative",
    ["V"],
    [
        FunctionType([ArrayType("V")], FloatType),  // objective: Array<V> -> Float
        ArrayType(ArrayType("V")),                   // parameter_spaces: per-element candidates
        IterativeConfigType,                         // config
    ],
    IterativeResultType
);

// ============================================================================
// Grouped Export
// ============================================================================

/**
 * Type definitions for iterative optimization.
 */
export const OptimizationTypes = {
    /** Initial value strategy variant */
    InitialStrategyType,
    /** Evaluation order variant */
    EvaluationOrderType,
    /** Configuration type */
    ConfigType: IterativeConfigType,
    /** Result type (with "V" placeholder for value type) */
    ResultType: IterativeResultType,
} as const;

/**
 * Iterative coordinate descent optimization for discrete combinatorial problems.
 *
 * Maximizes an objective by independently optimizing each element of a
 * parameter array over its candidate values. Supports multi-start sampling
 * for better exploration.
 *
 * Use cases:
 * - Task-worker assignment
 * - Scheduling and rostering
 * - Combinatorial selection problems
 * - Any discrete optimization with per-element candidate sets
 */
export const Optimization = {
    /**
     * Iterative optimization (generic over value type V).
     *
     * Call with type parameter array first, then arguments:
     * `Optimization.iterative([IntegerType], objective, spaces, config)`
     */
    iterative: optimization_iterative,

    /**
     * Type definitions for optimization functions.
     */
    Types: OptimizationTypes,
} as const;
